### ISSUE TITLE
Facing sql error while running import_views.py in chapter 7 (Async code)

### ISSUE THREAD

**Issue by gokulyc** [miguelgrinberg/retrofun#1]
Error Details:

sqlalchemy.dialects.postgresql.asyncpg.AsyncAdapt_asyncpg_dbapi.Error: <class 'asyncpg.exceptions.DataError'>: invalid input for query argument $2: UUID('d9e0b5fd-b794-49b3-947f-258bcf96a9... (expected str, got UUID)

I think there is issue with views.csv. Unable to Find & fix it.
[a.log](https://github.com/miguelgrinberg/retrofun/files/11506881/a.log)

**Comment by miguelgrinberg** [comment]
What is the output of `pip freeze`?
Were you able to import the orders data? Because that also uses UUIDs.

**Comment by gokulyc** [comment]
$ pipenv requirements > requirements.txt

[requirements.txt](https://github.com/miguelgrinberg/retrofun/files/11507951/requirements.txt)

All other scripts worked.

**Comment by miguelgrinberg** [comment]
Actually, yeah, I do get the error here as well. Let me debug this a bit. Will give you an update as soon as I can.

**Comment by miguelgrinberg** [comment]
Well, it appears this is a bug, but I'm still not sure if it is a SQLAlchemy bug or asyncpg bug.

This is what happens:

1. you create an instance of a model that has a UUID primary key
2. you create an instance of another model that has a foreign key pointing to the first model, and initialize it through a relationship attribute
3. when the session is committed, SQLAlchemy is supposed to flush the object in (1), obtain the assigned primary key from it, then write it in the foreign key column for the object in (2), and then commit the whole thing.

Step 3 appears to break when using asyncpg, it seems someone is expecting this foreign key is a string and not a UUID. This is fairly involved, so I will continue debugging to figure out who I need to send a bug report to.

**Comment by miguelgrinberg** [comment]
A workaround that allows this script to work is to add an explicit `flush()` for every `BlogView` model that is added to the session:

```python
                view = BlogView(
                    article=article,
                    session=blog_session,
                    timestamp=datetime.strptime(
                        row['timestamp'], '%Y-%m-%d %H:%M:%S'),
                )
                session.add(view)
                await session.flush()    # <--- add this statement
```

But I think this should not be necessary, so I will keep debugging to see why the implicit flushes do not work. But at least you shouldn't be blocked anymore if you want to continue with the book.

**Comment by gokulyc** [comment]
I created Session class with below snippet:

```python
Session = async_sessionmaker(engine, expire_on_commit=False, autoflush=False)
``` 

1. Here, my question is if autoflush=False, We are running session.flush() explicitly or this is special case due to some bug. Other import scripts doesn't have session.flush(). 

2. If autoflush=False, How come we are not using session.flush() for all scripts.

**Comment by miguelgrinberg** [comment]
As you've seen in the book, I mention `autoflush=False` as one possible option to use, but this isn't an option that I personally like or recommend. Without the automatic flushes some queries may appear to return inconsistent results.

To answer your questions:

1. The explicit `flush()` call that I suggested was while using `autoflush=True`, which is the default and also my recommendation. I have not tested the import script with autoflush disabled. The outcome may be different with autoflush disabled.
2. The `autoflush=False` setting does not mean that all flushes are disabled. In particular, a `commit()` call always implies a `flush()` call as well, regardless of how the `autoflush` option is set.

Automatic flushing is a specific behavior in SQLAlchemy that ensures that changes that are accumulated in your database session and aren't committed yet are included in any queries you issue in that same session. Let me see if I can make this more clear with an example:

- Let's say you create a customer with name `John` and add it to the session.
- Now within that session, you issue a query for all customers with names that start with `J`.
  - With `autoflush=True`, SQLAlchemy will flush the session right before issuing the query. This means that a database transaction will be created with this new customer. Then the query will be executed in the context of this transaction, and the query results will include `John`, along with any other customers that start with `J`.
  - With `autoflush=False`, SQLAlchemy will run the query directly. Since the `John` user is only known in the context of the SQLAlchemy session, the query results will not include it, since the database does not know it exists yet (it will know about it when the session is committed, which will also include a flush).

I hope this makes it more clear. To avoid this weird situation in which results that you expect to be included are not, I do not recommend using `autoflush=False`.

**Comment by gokulyc** [comment]
Thank you for clear & quick explanation.

**Comment by miguelgrinberg** [comment]
Okay, some more progress. This is a SQLAlchemy bug. I've found that the issue occurs when two or more entities with UUID columns are flushed together when using the asyncpg driver. An explicit flush after every instance is added works around the issue, as every one of those flushes will have only one entity pending flush.

A similar issue was reported for columns of type `LargeBinary` (see https://github.com/sqlalchemy/sqlalchemy/issues/9739) and the problem was addressed. I'm going to look through that fix and see if I can come up with a similar solution for UUIDs.
